bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Observation.scala

```diff
@@ -45,60 +52,89 @@ class Observation private (val blueprint:Option[BlueprintBase],
     case None => realTarget
   }
 
+  // REL-2985 requires special handling of the times.
+  // 2017A and earlier: intTimeParam will be None, and calculatedTimesParam will be defined with progTime.
+  // If a blueprint exists - which it should - we can calculate the intTime from the progTime and then the rest
+  // of the observation times from the intTime.
+  val intTime: Option[TimeAmount] = {
+    if (intTimeParam.nonEmpty) intTimeParam
+    else for {
+      b <- blueprint
+      c <- Overheads(b)
+      p <- progTimeParam
+    } yield c.intTimeFromProgTime(p)
+  }
+
+  // 2017B and later: if the blueprint and the intTime are defined, recalculate the calculatedTimes.
+  val calculatedTimes: Option[ObservationTimes] =  for {
+    b <- blueprint
+    c <- Overheads(b)
+    t <- intTime
+  } yield c.calculate(t)
+
+  def progTime:  Option[TimeAmount] = calculatedTimes.map(_.progTime)
+  def partTime:  Option[TimeAmount] = calculatedTimes.map(_.partTime)
+  def totalTime: Option[TimeAmount] = calculatedTimes.map(_.totalTime)
+
   def copy(blueprint:Option[BlueprintBase] = blueprint,
            condition:Option[Condition] = condition,
            target:Option[Target] = target,
-           time:Option[TimeAmount] = time,
+           intTime:Option[TimeAmount] = intTime,
+           calculatedTimes:Option[ObservationTimes] = calculatedTimes,
            band:Band = band,
            meta:Option[ObservationMeta] = None, // metadata resets to None on copy, by default
            enabled:Boolean = enabled) =
-    new Observation(blueprint, condition, target, time, band, meta, enabled)
+    new Observation(blueprint, condition, target, intTime, calculatedTimes.map(_.progTime), band, meta, enabled)
 
   def this(m:M.Observation) = this (
-    Option(m.getBlueprint).map(BlueprintBase(_)),
-    Option(m.getCondition).map(Condition(_)),
-    Option(m.getTarget).map(Target(_)),
-    Option(m.getTime).map(TimeAmount(_)),
-    Option(m.getBand).getOrElse(M.Band.BAND_1_2),
-    Option(m.getMeta).map(ObservationMeta(_)),
-    m.isEnabled
+      Option(m.getBlueprint).map(BlueprintBase(_)),
+      Option(m.getCondition).map(Condition(_)),
+      Option(m.getTarget).map(Target(_)),
+      Option(m.getIntTime).map(TimeAmount(_)),
+      Option(m.getProgTime).map(TimeAmount(_)),
```

Comment:
Carlos was suggesting pattern matching to make it more readable:
`(Option(m.getIntTime), Option(m.getProgTime)) match { ... }`
which is definitely better, I think, but there has to be some scalaz way, I would think, to do this more cleanly?